features/324-config/index.md

```diff
@@ -0,0 +1,56 @@
+# Revamp TenantConfiguration
+
+## Background
+
+TenantConfiguration serves the following purposes.
+
+- It stores non-editable app configuration, for example `APP_NAME`, `DATABASE_URL`.
+- It stores user configuration, for example `API_KEY`, `MASTER_KEY`, `CORS_HOST`.
+- It can be serialized into msgpack and JSON. The msgpack format is injected into HTTP header by the gateway and further consumed by gears. The JSON format is directly stored into the database.
+
+## Problems
+
+### Mixed app configuration and user configuration
+
+This is confusing and error-prone because we could accidentally allow user to modify app configuration.
+
+### Stored Derivable data
+
+It stores derivable data like `APP_NAME` where the source of truth is `app.name`.
+
+### Non user-facing
+
+The organization is not user-facing, for example there is `SOO_SETTING` and `SSO_CONFIGS`.
+
+### Non environment variable friendly name
+
+`_` is used in key name as aesthetic separator. This introduces ambiguity when we need to support loading configuration from environment variable where `_` is used to denote structure.
+
+### Stored deployment-specific data
+
+`HOOKS` is an example. Hooks should be derived from the incoming request. When we support serving multiple active deployments, the deployment-specific hooks should be invoked.
+
+Suppose we have two deployments
+
+- `https://tag1.myapp.skygearapis.com`
+- `https://tag2.myapp.skygearapis.com`
+
+If the signup request is `https://tag1.myapp.skygearapis.com/_auth/signup`, the hooks configuration injected by the gatway for the auth gear should be specific to the `tag1`.
+
+## Proposed solution
+
+The new TenantConfiguration is solely for gear consumption. When it is injected by the gateway, it is msgpack serialized. It can also be deserialized from environment variable in case the gear is running in standalone mode.
+
+The old TenantConfiguration will be splitted into AppConfiguration and UserConfiguration. AppConfiguration is used to store non-editable configuration while UserConfiguration is intended to be edited by the developer.
```

Comment:
I suppose user configuration would be associated with a deployment, instead of an app?

Reply:
The current design is associated with the app.

Reply:
How would the solution address the 'Stored deployment-specific data' problem you mentioned?

Reply:
The skycli command `skycli app update-user-config` does not take deployment id as argument so logically the config should apply to the app, not a specific deployment. However, hooks are configured in `skygear.yaml`, so they must be deployment-specific.